src/main/groovy/com/bugsnag/android/gradle/BugsnagPlugin.groovy

```diff
@@ -36,134 +43,133 @@ class BugsnagPlugin implements Plugin<Project> {
                 throw new IllegalStateException('Must apply \'com.android.application\' first!')
             }
 
-            // Create tasks for each Build Variant
-            // https://sites.google.com/a/android.com/tools/tech-docs/new-build-system/user-guide#TOC-Build-Variants
             project.android.applicationVariants.all { ApplicationVariant variant ->
-                if (hasDisabledBugsnag(variant)) {
-                    return
-                }
-
-                variant.outputs.all { output ->
-                    def usedOutput = variant.outputs.first()
-                    setupProguardAutoConfig(project, variant, usedOutput)
-                    setupManifestUuidTask(project, variant, usedOutput)
-                    setupMappingFileUpload(project, variant, usedOutput)
-                    setupNdkMappingFileUpload(project, variant, usedOutput)
-                }
+                applyBugsnagToVariant(variant, project)
             }
         }
     }
 
-    private static void setupMappingFileUpload(Project project, ApplicationVariant variant,  BaseVariantOutput output) {
-        // The Android build system supports creating multiple APKs
-        // per Build Variant (Variant Outputs):
-        // https://sites.google.com/a/android.com/tools/tech-docs/new-build-system/user-guide/apk-splits
-        //
-        // Variant Outputs share most tasks so we only need to attach
-        // Bugsnag tasks to the first output.
-        def variantName = variant.name.capitalize()
-
-        // Create a Bugsnag task to upload proguard mapping file
-        def uploadTaskClass = BugsnagUploadProguardTask
-        def uploadTask = project.tasks.create("uploadBugsnag${ variantName}Mapping", uploadTaskClass)
-        uploadTask.partName = isJackEnabled(project, variant) ? "jack" : "proguard"
+    /**
+     * Create tasks for each Build Variant
+     * See https://sites.google.com/a/android.com/tools/tech-docs/new-build-system/user-guide#TOC-Build-Variants
+     */
+    private void applyBugsnagToVariant(BaseVariant variant, Project project) {
+        if (hasDisabledBugsnag(variant)) {
+            return
+        }
 
-        uploadTask.group = GROUP_NAME
-        uploadTask.output = output
-        uploadTask.variant = variant
-        uploadTask.applicationId = variant.applicationId
-        uploadTask.mustRunAfter output.assemble
+        // only need to be run once per variant
+        setupProguardAutoConfig(project, variant)
+        setupSplitsDiscovery(project, variant)
 
-        if (project.bugsnag.autoUpload) {
-            project.tasks.findByName("package${variantName}").finalizedBy {uploadTask}
+        // need to be run for each output
+        variant.outputs.all { output ->
+            if (!output.name.toLowerCase().endsWith("debug")) {
```

Comment:
Is upload during debug something which would ever be needed/able to be enabled?

Reply:
I suppose if proguard were enabled in a debug build, this would produce a mapping file that could be uploaded. This doesn't seem like a common case as the functionality is disabled by default, and the typical user will probably install a debuggable APK multiple times each day.

Another thing to consider if we got rid of this check is that the debug upload tasks would currently log out a warning that a mapping file couldn't be found - we could disable this behaviour for debug builds. What are your thoughts on the best approach to take?

Reply:
It could be behind a (default disabled) option, like `uploadDebugBuildMappings`. It sounds like it makes sense to skip but we could still support people with an unusual configuration.

Reply:
Makes sense - I've added that flag into the plugin extension.